### PR TITLE
- coop scheduling extension

### PR DESCRIPTION
**Pull request by ezolenko**

### CODE REVIEW

File: dist/index.d.ts [hunk 1]
```diff
@@ -54,8 +56,24 @@ interface IPosisProcessRegistry {
 	// if your bundle consists of several programs you can pretend that we have a VFS: "ANI/MyBundle/BundledProgram1"
 	register(imageName: string, constructor: new (context: IPosisProcessContext) => IPosisProcess): boolean;
 }
+interface IPosisCooperativeSchedulingV1 {
```
Comment (anisoptera) [inline dist/index.d.ts]: Can we just remove V1? A stub implementation of yield is trivial. An empty function fulfills the requirement.

File: dist/index.d.ts [hunk 2]
```diff
@@ -70,7 +88,22 @@ declare const enum EPosisSpawnStatus {
...
+        //     Host must guarantee that the spawning creep will have one of provided bodies
+        //     Which body to spawn is up to host to select based on its own logic
+        body: string[][], 
+        //   - 'priority' is spawn priority in range -1000 (the highest) to 1000 (the lowest)
+        //     Used as a hint for host's logic. Host may (but not guarantee) spawn creeps in priority order
+        priority?: number, 
+        //   - 'pid' is id of the process which is spawned creep associated to
+        //     Used as a hint for host's logic. Host may (but not guarantee) consider currently spawned creeps
+        //     detached, may (but not guarantee) remove scheduled creeps from queue on process termination
+        pid?: PosisPID }): string;
```
Comment (anisoptera) [inline dist/index.d.ts]: This doesn't belong in this PR :)

File: dist/index.d.ts [hunk 3]
```diff
@@ -54,8 +56,24 @@ interface IPosisProcessRegistry {
...
+    // CPU budget scheduler allocated to this process. 
+    readonly budget: number;
+    // Process can call yield with a callback when it is ready to give up for the tick, but can continue if CPU is available. 
+    // Call will either return, indicating there is spare CPU, or callback will be called and yield will not return, ending execution for current tick.
+    // Use callback to do last minute tasks like saving current state, etc.
+    // The call will throw, so avoid catching generic exceptions around it.
+    yield(cb: () => void): void;
+}
 declare const enum EPosisSpawnStatus {
-    UNKNOWN = -2,
```
Comment (anisoptera) [inline dist/index.d.ts]: ditto this one

**Review by anisoptera — CHANGES_REQUESTED** [review]

File: dist/index.d.ts (hunk 2 above)
Comment (ezolenko) [reply dist/index.d.ts]: It is not in this PR :), it is in master already. I blame github.

File: dist/index.d.ts (hunk 1 above)
Comment (ezolenko) [reply dist/index.d.ts]: Unimplemented yield is worse than no yield. Process can't detect when it needs to fall back to manual yield by doing `if (!budgetLeft) return`

Ah, you can make optional functions in ts. I wanted to avoid this symtax `yield?: (cb: () => void) => void`. :). I'll change it.

File: dist/index.d.ts (hunk 2 above)
Comment (AlinaNova21) [reply dist/index.d.ts]: and this is why my repos never contain build or dist folders ;)

**Review by AlinaNova21 — APPROVED** [review]

**Review by anisoptera — APPROVED** [review]

**Comment by RiftLurker** [comment]
Instead of simulating the behaviour of ES2015 generator functions with our own yield solution with a "throw when done" approach, it might be a good idea to just use them directly. We've already discussed this in slack but I'm not sure if we came to a final decision.

I've modified the [most recent version]((https://screeps.slack.com/files/ags131/F60U3EC7M/Untitled.js)) I found, renaming some of the variables and adding more stuff to the process:
```typescript
// spec
interface IPosisCooperativeScheduling {
  readonly used: number;
  readonly budget: number;
  wrap?(makeIterator: () => IterableIterator<void | (() => void)>): void;
}

// example process
class CooperativeProcess implements IPosisProcess {

  context: IPosisProcessContext;

  constructor(context: IPosisProcessContext) {
    this.context = context;
  }

  run() {
    const coop = this.context.queryPosisInterface('coop');
    if(!coop){
      let now = Game.cpu.getUsed()
      coop = { // Shim it!
        budget: Game.cpu.limit - now,
        get used(){ return Game.cpu.getUsed() - now }
      }
    }
    if(!coop.wrap){
      coop.wrap = function(makeIterator){
        const iterator = makeIterator();
        let result = undefined;
        while(coop.used < coop.budget) {
          if (result && result.done) {
            break;
          }
          result = iterator.next();
        }
        if (result && typeof result.value == 'function') {
          return result.value();
        }
      }
    }
    coop.wrap(()=>this.runCoop())
    
  }

  * runCoop() {
    // do expensive stuff
    yield; // give kernel option to cancel
    // do more stuff
    yield () => { // give kernel another option to cancel, this time with shutdown function
      // save results
    };
    // and some pretty logging if you want to
  }
```

If we agree on this I can create a PR to @ezolenko's fork to add this.

**Comment by AlinaNova21** [comment]
@PostCrafter you pass in as `makeGenerator` then use `makeIterator` ;) Otherwise, looks good to me

**Comment by RiftLurker** [comment]
@ags131 Thanks, fixed it.

**Comment by ezolenko** [comment]
Ok, changed to PostCrafter's version
